[WebGPU] Implement PresentationContextCoreAnimation

litherum opened the pull request:
#### 626bc3146478155fc72b5c49b5282e5032a864d1
<pre>
[WebGPU] Implement PresentationContextCoreAnimation
<a href="https://bugs.webkit.org/show_bug.cgi?id=250775">https://bugs.webkit.org/show_bug.cgi?id=250775</a>
rdar://104388628

Reviewed by Dean Jackson.

This is enough for us to be able to use WebGPUPlayground&apos;s AppPlayground for
WebGPU development.

No tests because there is no behavior change (from WebKit&apos;s perspective).

* Source/WebGPU/WebGPU/HardwareCapabilities.h:
* Source/WebGPU/WebGPU/HardwareCapabilities.mm:
(WebGPU::apple3):
(WebGPU::apple4):
(WebGPU::apple5):
(WebGPU::apple6):
(WebGPU::apple7):
(WebGPU::mac2):
(WebGPU::mergeBaseCapabilities):
* Source/WebGPU/WebGPU/PresentationContextCoreAnimation.h:
(WebGPU::PresentationContextCoreAnimation::Configuration::Configuration):
* Source/WebGPU/WebGPU/PresentationContextCoreAnimation.mm:
(WebGPU::layerFromSurfaceDescriptor):
(WebGPU::PresentationContextCoreAnimation::PresentationContextCoreAnimation):
(WebGPU::PresentationContextCoreAnimation::configure):
(WebGPU::PresentationContextCoreAnimation::Configuration::generateCurrentFrameState):
(WebGPU::PresentationContextCoreAnimation::present):
(WebGPU::PresentationContextCoreAnimation::getCurrentTextureView):
* Source/WebGPU/WebGPU/PresentationContextIOSurface.mm:
* Tools/WebGPUPlayground/AppPlayground/WebGPUCoordinator.m:
(-[WebGPUCoordinator mtkView:drawableSizeWillChange:]):

Canonical link: <a href="https://commits.webkit.org/259055@main">https://commits.webkit.org/259055@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a6c825bb30ce83b43808d59e5b6cf106742b84c9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103784 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12901 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36726 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113008 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/173329 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107733 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13922 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3795 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96033 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/112129 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109555 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/10735 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/38451 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/92566 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/macOS-BigSur-Release-WK1-Tests-EWS "Waiting to run tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/80102 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6258 "Built successfully") | [⏳ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/macOS-BigSur-Release-WK2-Tests-EWS "Waiting to run tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6427 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/3331 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12410 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46315 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6229 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8194 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/HardwareCapabilities.h:
```diff
@@ -39,6 +39,7 @@ struct HardwareCapabilities {
         bool supportsNonPrivateDepthStencilTextures { false };
         id<MTLCounterSet> timestampCounterSet { nil };
         id<MTLCounterSet> statisticCounterSet { nil };
+        bool canPresent10A2PixelFormats { false };
```
What does `10A2` mean in this context? Is it BGR10A2 or RGB10A2 or both?

litherum replied:
Both, as per https://developer.apple.com/documentation/quartzcore/cametallayer/1478155-pixelformat?language=objc

mwyrzykowski commented on Source/WebGPU/WebGPU/PresentationContextCoreAnimation.mm:
```diff
@@ -27,30 +27,100 @@
+    case WGPUTextureFormat_BGRA8UnormSrgb:
+    case WGPUTextureFormat_RGBA16Float:
+        break;
+    case WGPUTextureFormat_RGB10A2Unorm:
```
little unfortunate we do not have `BGR10A2` formats should we request that?

litherum replied:
(Conversation continued offline)

mwyrzykowski commented on Source/WebGPU/WebGPU/HardwareCapabilities.mm:
```diff
@@ -96,6 +96,7 @@ static HardwareCapabilities apple3(id<MTLDevice> device)
     auto baseCapabilities = WebGPU::baseCapabilities(device);
 
     baseCapabilities.supportsNonPrivateDepthStencilTextures = true;
+    baseCapabilities.canPresent10A2PixelFormats = false;
```
where does this come from? I thought all devices since Apple3 supported RGB10A2: https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf

litherum replied:
https://developer.apple.com/documentation/quartzcore/cametallayer/1478155-pixelformat?language=objc

mwyrzykowski reviewed (changes requested):
I don't think canPresentRGB10A2PixelFormats should be placed in hardware capabilities, it seems to be a CAMetalLayer thing

litherum committed:
[WebGPU] Implement PresentationContextCoreAnimation
https://bugs.webkit.org/show_bug.cgi?id=250775
rdar://104388628

Reviewed by Dean Jackson.

This is enough for us to be able to use WebGPUPlayground's AppPlayground for
WebGPU development.

No tests because there is no behavior change (from WebKit's perspective).

* Source/WebGPU/WebGPU/HardwareCapabilities.h:
* Source/WebGPU/WebGPU/HardwareCapabilities.mm:
(WebGPU::apple3):
(WebGPU::apple4):
(WebGPU::apple5):
(WebGPU::apple6):
(WebGPU::apple7):
(WebGPU::mac2):
(WebGPU::mergeBaseCapabilities):
* Source/WebGPU/WebGPU/PresentationContextCoreAnimation.h:
(WebGPU::PresentationContextCoreAnimation::Configuration::Configuration):
* Source/WebGPU/WebGPU/PresentationContextCoreAnimation.mm:
(WebGPU::layerFromSurfaceDescriptor):
(WebGPU::PresentationContextCoreAnimation::PresentationContextCoreAnimation):
(WebGPU::PresentationContextCoreAnimation::configure):
(WebGPU::PresentationContextCoreAnimation::Configuration::generateCurrentFrameState):
(WebGPU::PresentationContextCoreAnimation::present):
(WebGPU::PresentationContextCoreAnimation::getCurrentTextureView):
* Source/WebGPU/WebGPU/PresentationContextIOSurface.mm:
* Tools/WebGPUPlayground/AppPlayground/WebGPUCoordinator.m:
(-[WebGPUCoordinator mtkView:drawableSizeWillChange:]):

Canonical link: https://commits.webkit.org/259055@main
